File: binary-trees/n-ary-tree-preorder-traversal/n-ary-tree-preorder-traversal.js
function Node(val, children) {
  this.val = val;
  this.children = children;
}

/**
 * @param {Node} root
 * @return {number[]}
 */
function preorder(root) {
  const list = [];
  traverse(root, list);
  return list;
}

/**
 *
 * @param {Node} root
 * @param {number[]} list
 * @returns
 */
function traverse(root, list) {
  if (root === null) {
    return;
  }

  list.push(root.val);

  for (let node of root.children) {
    traverse(node, list);
  }
}

File: binary-trees/n-ary-tree-preorder-traversal/n-ary-tree-preorder-traversal.go
package narytreepreordertraversal

type Node struct {
	Val      int
	Children []*Node
}

func preorder(root *Node) []int {
	list := new([]int)
	traverse(root, list)
	return *list
}

func traverse(root *Node, list *[]int) {
	if root == nil {
		return
	}

	*list = append(*list, root.Val)

	for _, node := range root.Children {
		traverse(node, list)
	}
}
